### DISCUSSION TITLE
Am I able to implement a background event loop that runs forever using this library

### DISCUSSION THREAD

**Discussion by stan-stately** [socketry/async#365] [General]
im trying to write an auth token provider. This provider needs to fetch tokens in the background asynchronously. A minimal repro of what im trying to do is this:
```ruby
# setup a global scheduler for the process
scheduler = Async::Scheduler.new
Fiber.set_scheduler(scheduler)       

class Test
  def initialize
    @res = 0
    Async do
      increment_slowly
      puts "background done: #{@res}"
    end
  end

  def get_sync
    Sync do
      increment_slowly.wait
    end
  end

  def increment_slowly
    Async do
      puts "incrementing"
      sleep(1)
      @res += 1
      @res
    end
  end

  attr_reader :res
end
```

I would expect the following test to pass but `@res` is always 0:
```ruby
RSpec.describe "test" do
  it "works" do
    test = Test.new
    sleep(2)
    # background operation should have finished by now
    expect(test.res).to eql(1)
    expect(test.get_sync).to eql(2)
  end
end
```

Someone asked a similar question here: https://github.com/socketry/async/discussions/268 but their solution seems over complicated.

What do you think?

**Comment by stan-stately** [comment]
I should note that I don't want to expose the fact that this provider uses async internally to consumers of my library

**Reply by ioquatix** [reply]
You should look at `async-actor` then: https://github.com/socketry/async-actor

You can hide whatever you want inside an event driven actor with a synchronous interface.

**Reply by stan-stately** [reply]
thanks. that worked really well

**Reply by ioquatix** [reply]
Is your library open source?

**Reply by stan-stately** [reply]
it is actually: https://github.com/StatelyCloud/ruby-sdk/blob/main/lib/common/auth/auth0_token_provider.rb

**Reply by ioquatix** [reply]
A few thoughts - of no particular significance.

You might like to make `#close` a little more robust, e.g.

```ruby
def close
  if actor = @actor
    @actor = nil
    actor.close
  end
end
```

You might like to rename `@actor.init` to `@actor.start` or something that indicates that it's starting the background refresh operation.

I don't think your implementation of `refresh_token` returns a string, but instead returns a Task. In fact, I'd recommend adding `return nil` at the end of that method.

`class Actor` seems like it could be named something more specific to the task at hand, e.g. `BackgroundTokenRefresher` or something. The fact that you wrap it  in an actor shouldn't really impact the implementation (you should in theory be able to directly invoke this code from within the event loop).

Instead of having `refresh_token_impl` call `refresh_token` recursively, I wonder if you can just do it in a loop?
